crates/core/src/sql/execute.rs

```diff
@@ -56,39 +74,61 @@ pub(crate) fn collect_result(result: &mut Vec<MemTable>, r: CodeResult) -> Resul
 /// Evaluates `ast` and accordingly triggers mutable or read tx to execute
 ///
 /// Also, in case the execution takes more than x, log it as `slow query`
-pub fn execute_sql(db: &RelationalDB, sql: &str, ast: Vec<CrudExpr>, auth: AuthCtx) -> Result<Vec<MemTable>, DBError> {
+pub fn execute_sql(db: &RelationalDB, sql: &str, ast: Vec<CrudExpr>, auth: AuthCtx, subs: &ModuleSubscriptions) -> Result<Vec<MemTable>, DBError> {
     let total = ast.len();
+    let reads = CrudExpr::is_reads(&ast);
     let ctx = ExecutionContext::sql(db.address(), db.read_config().slow_query);
-    let mut result = Vec::with_capacity(total);
     let sources = [].into();
     let slow_logger = SlowQueryLogger::query(&ctx, sql);
-
-    match CrudExpr::is_reads(&ast) {
-        false => db.with_auto_commit(&ctx, |mut_tx| {
+    let q = Expr::Block(ast.into_iter().map(|x| Expr::Crud(Box::new(x))).collect());
+    let (code, subscriptions): (Result<_, DBError>, _) = if reads {
+        (db.with_auto_commit(&ctx, |mut_tx| {
             let mut tx: TxMode = mut_tx.into();
-            let q = Expr::Block(ast.into_iter().map(|x| Expr::Crud(Box::new(x))).collect());
             let p = &mut DbProgram::new(&ctx, db, &mut tx, auth);
             // SQL queries can never reference `MemTable`s, so pass an empty `SourceSet`.
-            collect_result(&mut result, run_ast(p, q, sources).into())
-        }),
-        true => db.with_read_only(&ctx, |tx| {
+            Ok(run_ast(p, q, sources))
+        }), Some(subs.subscriptions.read()))
+    }
+    else {
+        (db.with_read_only(&ctx, |tx| {
             let mut tx = TxMode::Tx(tx);
-            let q = Expr::Block(ast.into_iter().map(|x| Expr::Crud(Box::new(x))).collect());
             let p = &mut DbProgram::new(&ctx, db, &mut tx, auth);
             // SQL queries can never reference `MemTable`s, so pass an empty `SourceSet`.
-            collect_result(&mut result, run_ast(p, q, sources).into())
-        }),
-    }?;
+            Ok(run_ast(p, q, sources))
+        }), None)
+    };
     slow_logger.log();
 
-    Ok(result)
+    let mut tables = Vec::with_capacity(total);
+    let mut updates = Vec::with_capacity(total);
+    collect_result(&mut tables, &mut updates, code?.into())?;
+
+    if !reads && !updates.is_empty() {
+        let event = ModuleEvent {
```

Comment:
Left most of the values to defaults here, assuming these relate to reflecting the state of a reducer/client which aren't present in this case.

Only handling the committed case, as collect_result short-circuits the function on error.